File: casper/src/test/scala/coop/rchain/casper/engine/InitializingSpec.scala
package coop.rchain.casper.engine

import cats.effect.Concurrent
import cats.syntax.all._
import com.google.protobuf.ByteString
import coop.rchain.casper._
import coop.rchain.casper.protocol._
import coop.rchain.catscontrib.TaskContrib._
import coop.rchain.catscontrib.ski._
import coop.rchain.comm.rp.ProtocolHelper._
import coop.rchain.crypto.hash.Blake2b256
import coop.rchain.crypto.signatures.Secp256k1
import coop.rchain.rholang.interpreter.RhoRuntime
import coop.rchain.rspace.hashing.Blake2b256Hash
import coop.rchain.rspace.syntax._
import coop.rchain.shared.{Cell, EventPublisher}
import fs2.concurrent.Queue
import monix.eval.Task
import org.scalatest.{BeforeAndAfterEach, WordSpec}

import scala.concurrent.duration._

class InitializingSpec extends WordSpec with BeforeAndAfterEach {
  implicit val eventBus = EventPublisher.noop[Task]

  val fixture = Setup()
  import fixture._

  override def beforeEach(): Unit =
    transportLayer.setResponses(_ => p => Right(()))

  override def afterEach(): Unit =
    transportLayer.reset()

  "Initializing state" should {
    // This test verifies the fix for the race condition where a slow validator
    // misses the ApprovedBlock during genesis ceremony. The fix is to proactively
    // request the ApprovedBlock when entering Initializing state, rather than
    // waiting for it to arrive (which may never happen if it was already broadcast
    // and dropped while the node was still in GenesisValidator state).
    "proactively request ApprovedBlock on init" in {
      implicit val engineCell = Cell.unsafe[Task, Engine[Task]](Engine.noop)

      val blockResponseQueue = Queue.unbounded[Task, BlockMessage].runSyncUnsafe()
      val stateResponseQueue = Queue.unbounded[Task, StoreItemsMessage].runSyncUnsafe()

      val initializingEngine =
        new Initializing[Task](
          fixture.blockProcessingQueue,
          fixture.blockProcessingState,
          fixture.casperShardConf,
          Some(validatorId),
          Task.unit, // theInit
          blockResponseQueue,
          stateResponseQueue,
          trimState = true,
          disableStateExporter = false,
          onBlockFinalized = (_: String) => Task.unit
        )

      // Clear any previous transport requests and set up responses
      transportLayer.reset()
      transportLayer.setResponses(_ => _ => Right(()))

      val expectedContent = ApprovedBlockRequest("", trimState = true).toProto.toByteString

      // Poll until the request appears (sendWithRetry spawns a fiber)
      def pollForRequest(maxAttempts: Int, interval: FiniteDuration): Task[Unit] =
        Task.defer {
          val requests = transportLayer.requests
          if (requests.exists(_.msg.message.packet.exists(_.content == expectedContent))) {
            Task.unit
          } else if (maxAttempts <= 0) {
            Task.raiseError(
              new AssertionError(
                s"Initializing.init should send ApprovedBlockRequest within timeout. Requests sent: ${requests
                  .map(_.msg)}"
              )
            )
          } else {
            Task.sleep(interval) >> pollForRequest(maxAttempts - 1, interval)
          }
        }

      val test = for {
        // Call init - this should proactively request ApprovedBlock
        _ <- initializingEngine.init
        // Poll for the request with timeout (50 attempts * 10ms = 500ms max)
        _ <- pollForRequest(maxAttempts = 50, interval = 10.millis)
      } yield ()

      test.unsafeRunSync
    }

    "make a transition to Running once ApprovedBlock has been received" in {
      val theInit = Task.unit

      implicit val engineCell = Cell.unsafe[Task, Engine[Task]](Engine.noop)

      val blockResponseQueue = Queue.unbounded[Task, BlockMessage].runSyncUnsafe()
      val stateResponseQueue = Queue.unbounded[Task, StoreItemsMessage].runSyncUnsafe()

      // interval and duration don't really matter since we don't require and signs from validators
      val initializingEngine =
        new Initializing[Task](
          fixture.blockProcessingQueue,
          fixture.blockProcessingState,
          fixture.casperShardConf,
          Some(validatorId),
          theInit,
          blockResponseQueue,
          stateResponseQueue,
          trimState = true,
          disableStateExporter = false,
          onBlockFinalized = (_: String) => Task.unit
        )

      val approvedBlockCandidate = ApprovedBlockCandidate(block = genesis, requiredSigs = 0)

      val approvedBlock: ApprovedBlock = ApprovedBlock(
        candidate = approvedBlockCandidate,
        sigs = List(
          Signature(
            ByteString.copyFrom(validatorPk.bytes),
            "secp256k1",
            ByteString.copyFrom(
              Secp256k1
                .sign(Blake2b256.hash(approvedBlockCandidate.toProto.toByteArray), validatorSk)
            )
          )
        )
      )

      // Get exporter for genesis block
      val genesisExporter = exporter

      val chunkSize = LfsTupleSpaceRequester.pageSize

      // Export history and data from RSpace (genesis block)
      def genesisExport(startPath: Seq[(Blake2b256Hash, Option[Byte])]) =
        for {
          items <- genesisExporter.getHistoryAndData(
                    startPath,
                    skip = 0,
                    take = chunkSize
                    // ByteString.copyFrom
                  )
          // Exported history and data items
          (history, data) = items
        } yield (history.items, data.items, history.lastPath)

      val postStateHashBs = approvedBlock.candidate.block.body.state.postStateHash
      val postStateHash   = Blake2b256Hash.fromByteString(postStateHashBs)
      val startPath1      = Seq((postStateHash, none))

      // Get history and data items from genesis block
      val (historyItems1, dataItems1, lastPath1) = genesisExport(startPath1).runSyncUnsafe()
      val (historyItems2, dataItems2, lastPath2) = genesisExport(lastPath1).runSyncUnsafe()

      // Store request message
      val storeRequestMessage1 = StoreItemsMessageRequest(startPath1, 0, chunkSize)
      val storeRequestMessage2 = StoreItemsMessageRequest(lastPath1, 0, chunkSize)

      // Store response message
      val storeResponseMessage1 =
        StoreItemsMessage(startPath1, lastPath1, historyItems1, dataItems1)
      val storeResponseMessage2 =
        StoreItemsMessage(lastPath1, lastPath2, historyItems2, dataItems2)

      // Block request message
      val blockRequestMessage = BlockRequest(genesis.blockHash)

      // Send two response messages to signal the end
      val enqueueResponses = stateResponseQueue.enqueue1(storeResponseMessage1) *>
        stateResponseQueue.enqueue1(storeResponseMessage2) *>
        // Send block response
        blockResponseQueue.enqueue1(genesis)

      val expectedRequests = Seq(
        packet(local, networkId, storeRequestMessage1.toProto),
        packet(local, networkId, storeRequestMessage2.toProto),
        packet(local, networkId, blockRequestMessage.toProto),
        packet(local, networkId, ForkChoiceTipRequestProto())
      )

      val test = for {
        _ <- EngineCell[Task].set(initializingEngine)

        // Send responses with some delay because `Initializing.handle` is blocking until LFS is not received.
        _ <- Concurrent[Task].start(Task.sleep(5.second) >> enqueueResponses)

        // Handle approved block (it's blocking until responses are received)
        _ <- initializingEngine.handle(local, approvedBlock)

        engine          <- EngineCell[Task].read
        casperDefined   <- engine.withCasper(kp(true.pure[Task]), false.pure[Task])
        _               = assert(casperDefined)
        blockO          <- blockStore.get(genesis.blockHash)
        _               = assert(blockO.isDefined)
        _               = assert(blockO.contains(genesis))
        handlerInternal <- EngineCell[Task].read
        _               = assert(handlerInternal.isInstanceOf[Running[Task]])

        // Assert all expected messages were requested (may include retries due to backoff)
        messages = transportLayer.requests.map(_.msg).toSet
        _ = assert(
          expectedRequests.forall(messages.contains),
          s"Missing expected requests. Got: $messages"
        )
        _ = transportLayer.reset()

        // assert that we really serve last approved block
        lastApprovedBlockO <- LastApprovedBlock[Task].get
        _                  = assert(lastApprovedBlockO.isDefined)
        _ <- EngineCell[Task].read >>= (_.handle(
              local,
              ApprovedBlockRequest("test", trimState = false)
            ))
        // TODO: fix missing signature in received approved block
        _ = assert(
          transportLayer.requests
            .exists(_.msg.message.packet.get.content == approvedBlock.toProto.toByteString)
        )
      } yield ()

      test.unsafeRunSync
    }
  }

}

File: casper/tests/engine/initializing_spec.rs
// See casper/src/test/scala/coop/rchain/casper/engine/InitializingSpec.scala

use rspace_plus_plus::rspace::state::instances::rspace_exporter_store::RSpaceExporterStore;
use std::collections::HashSet;
use std::future::Future;
use std::pin::Pin;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc,
};
use tokio::sync::mpsc;

use crypto::rust::{
    hash::blake2b256::Blake2b256,
    signatures::{secp256k1::Secp256k1, signatures_alg::SignaturesAlg},
};
use models::casper::Signature;
use models::routing::protocol::Message as ProtocolMessage;
use models::rust::casper::protocol::casper_message::{
    ApprovedBlock, ApprovedBlockRequest, BlockMessage, BlockRequest, CasperMessage,
    StoreItemsMessage, StoreItemsMessageRequest,
};
use prost::bytes::Bytes;
use prost::Message;
use shared::rust::shared::f1r3fly_events::{EventPublisher, EventPublisherFactory};

use crate::engine::setup::TestFixture;
use casper::rust::engine::engine::transition_to_initializing;
use casper::rust::engine::engine_cell::EngineCell;
use casper::rust::engine::initializing::Initializing;
use casper::rust::engine::lfs_tuple_space_requester;

use casper::rust::errors::CasperError;
use comm::rust::rp::protocol_helper::packet_with_content;
use comm::rust::test_instances::TransportLayerStub;
use rspace_plus_plus::rspace::hashing::blake2b256_hash::Blake2b256Hash;
use rspace_plus_plus::rspace::state::exporters::rspace_exporter_items::RSpaceExporterItems;
use rspace_plus_plus::rspace::state::rspace_exporter::RSpaceExporter;
use shared::rust::ByteVector;

struct InitializingSpec;

impl InitializingSpec {
    fn event_bus() -> Box<dyn EventPublisher> {
        EventPublisherFactory::noop()
    }

    fn before_each(fixture: &TestFixture) {
        fixture
            .transport_layer
            .set_responses(|_peer, _protocol| Ok(()));
    }

    fn after_each(fixture: &TestFixture) {
        fixture.transport_layer.reset();
    }
    async fn make_transition_to_running_once_approved_block_received() {
        let _event_bus = Self::event_bus();

        let fixture = TestFixture::new().await;

        Self::before_each(&fixture);

        let the_init = Arc::new(|| {
            Box::pin(async { Ok(()) })
                as Pin<Box<dyn Future<Output = Result<(), CasperError>> + Send>>
        });

        let engine_cell = Arc::new(EngineCell::init());

        // interval and duration don't really matter since we don't require and signs from validators
        let initializing_engine =
            create_initializing_engine(&fixture, the_init, engine_cell.clone())
                .await
                .expect("Failed to create Initializing engine");

        let genesis = &fixture.genesis;
        let approved_block_candidate = fixture.approved_block_candidate.clone();
        let validator_sk = &fixture.validator_sk;
        let validator_pk = &fixture.validator_pk;

        let approved_block = {
            let candidate_proto = approved_block_candidate.clone().to_proto();
            let candidate_bytes = {
                let mut buf = Vec::new();
                Message::encode(&candidate_proto, &mut buf).expect("Failed to encode candidate");
                buf
            };
            let candidate_hash = Blake2b256::hash(candidate_bytes);
            let signature_bytes = Secp256k1.sign(&candidate_hash, &validator_sk.bytes);

            ApprovedBlock {
                candidate: approved_block_candidate,
                sigs: vec![Signature {
                    public_key: validator_pk.bytes.clone(),
                    algorithm: "secp256k1".to_string(),
                    sig: signature_bytes.into(),
                }],
            }
        };

        // Get exporter for genesis block
        // Note: Instead of default exported, we should use RSpaceExporterItems::get_history_and_data
        //let _genesis_exporter = &fixture.exporter;

        let chunk_size = lfs_tuple_space_requester::PAGE_SIZE;

        fn genesis_export(
            genesis_exporter: Arc<dyn RSpaceExporter>,
            start_path: Vec<(Blake2b256Hash, Option<u8>)>,
            exporter_params: &crate::engine::setup::ExporterParams,
        ) -> Result<
            (
                Vec<(Blake2b256Hash, ByteVector)>,
                Vec<(Blake2b256Hash, ByteVector)>,
                Vec<(Blake2b256Hash, Option<u8>)>,
            ),
            String,
        > {
            let (history_store_items, data_store_items) = RSpaceExporterItems::get_history_and_data(
                genesis_exporter,
                start_path,
                exporter_params.skip,
                exporter_params.take,
            );
            Ok((
                history_store_items.items,
                data_store_items.items,
                history_store_items.last_path,
            ))
        }

        let post_state_hash_bs = &approved_block.candidate.block.body.state.post_state_hash;
        let post_state_hash = Blake2b256Hash::from_bytes_prost(post_state_hash_bs);
        let start_path1 = vec![(post_state_hash, None::<u8>)];

        let rspace_store = &fixture.rspace_store;
        let genesis_exporter_impl = RSpaceExporterStore::create(
            rspace_store.history.clone(),
            rspace_store.cold.clone(),
            rspace_store.roots.clone(),
        );
        let genesis_exporter_arc = Arc::new(genesis_exporter_impl);

        // Build all StoreItems requests/responses dynamically until exporter indicates completion.
        // The number of chunks can change when genesis tuplespace shape changes.
        let mut store_request_messages = Vec::new();
        let mut store_response_messages = Vec::new();
        let mut next_start_path = start_path1.clone();
        let mut seen_paths = HashSet::new();

        loop {
            if !seen_paths.insert(next_start_path.clone()) {
                break;
            }

            let request = StoreItemsMessageRequest {
                start_path: next_start_path.clone(),
                skip: 0,
                take: chunk_size,
            };

            let (history_items, data_items, last_path) = genesis_export(
                genesis_exporter_arc.clone(),
                next_start_path.clone(),
                &fixture.exporter_params,
            )
            .expect("Failed to export history and data items");

            let response = StoreItemsMessage {
                start_path: next_start_path.clone(),
                last_path: last_path.clone(),
                history_items: history_items
                    .into_iter()
                    .map(|(hash, bytes)| (hash, Bytes::from(bytes)))
                    .collect(),
                data_items: data_items
                    .into_iter()
                    .map(|(hash, bytes)| (hash, Bytes::from(bytes)))
                    .collect(),
            };

            store_request_messages.push(request);
            store_response_messages.push(response);

            if last_path.is_empty() || last_path == next_start_path {
                break;
            }
            next_start_path = last_path;

            assert!(
                store_request_messages.len() < 1024,
                "Too many tuple-space chunks while preparing initializing_spec test"
            );
        }

        // Block request message
        let block_request_message = BlockRequest {
            hash: genesis.block_hash.clone(),
        };

        // Send two response messages to signal the end
        // Scala equivalent: stateResponseQueue.enqueue1(storeResponseMessage1) *>
        //                   stateResponseQueue.enqueue1(storeResponseMessage2) *>
        //                   blockResponseQueue.enqueue1(genesis)
        // IMPORTANT: Write directly to channels (NOT through handle()) like Scala test does
        let tuple_space_tx = initializing_engine
            .tuple_space_tx
            .lock()
            .unwrap()
            .as_ref()
            .unwrap()
            .clone();
        let block_message_tx = initializing_engine
            .block_message_tx
            .lock()
            .unwrap()
            .as_ref()
            .unwrap()
            .clone();

        let store_msgs_clone = store_response_messages.clone();
        let genesis_clone = genesis.clone();

        let enqueue_responses = async move {
            // Write directly to tuple space channel (equivalent to stateResponseQueue.enqueue1)
            for store_msg in store_msgs_clone {
                tuple_space_tx
                    .send(store_msg)
                    .await
                    .expect("Failed to enqueue tuple space response");
            }
            // Write directly to block message channel (equivalent to blockResponseQueue.enqueue1)
            block_message_tx
                .send(genesis_clone)
                .await
                .expect("Failed to enqueue block response");
        };

        let local_for_expected = fixture.local.clone();
        let mut expected_requests: Vec<_> = store_request_messages
            .iter()
            .map(|request| {
                packet_with_content(
                    &local_for_expected,
                    &fixture.network_id,
                    request.clone().to_proto(),
                )
            })
            .collect();
        expected_requests.push(packet_with_content(
            &local_for_expected,
            &fixture.network_id,
            block_request_message.to_proto(),
        ));
        expected_requests.push(packet_with_content(
            &local_for_expected,
            &fixture.network_id,
            models::casper::ForkChoiceTipRequestProto::default(),
        ));

        let test = async {
            engine_cell.set(initializing_engine.clone()).await;

            let enqueue_responses_with_delay = async move {
                tokio::time::sleep(tokio::time::Duration::from_secs(5)).await;
                enqueue_responses.await;
            };

            let approved_block_clone = approved_block.clone();
            let local_for_handle = fixture.local.clone();
            // Handle approved block (it's blocking until responses are received)
            let handle_fut = async {
                let engine = engine_cell.get().await;
                engine
                    .handle(
                        local_for_handle,
                        CasperMessage::ApprovedBlock(approved_block_clone),
                    )
                    .await
                    .expect("Failed to handle approved block");
            };
            let _ = tokio::join!(enqueue_responses_with_delay, handle_fut);

            let engine = engine_cell.get().await;

            let casper_defined = engine.with_casper().is_some();
            assert!(
                casper_defined,
                "Casper should be defined after handling approved block"
            );

            let block_option = fixture
                .block_store
                .get(&genesis.block_hash)
                .expect("Failed to get block from store");
            assert!(block_option.is_some(), "Block should be defined in store");
            assert_eq!(block_option.as_ref(), Some(genesis));

            let handler_internal = engine_cell.get().await;

            // We use with_casper().is_some() as a proxy: Running engines have casper, Initializing engines return None.
            // This is functionally equivalent since after transition_to_running(), only Running engines should be in the cell.
            assert!(
                handler_internal.with_casper().is_some(),
                "Engine should be Running (checked via casper presence)"
            );

            let requests = fixture.transport_layer.get_all_requests();
            // Assert requested messages for the state and fork choice tip
            assert_eq!(
                requests.len(),
                expected_requests.len(),
                "Transport layer should have received expected number of requests"
            );

            // Note: Since Protocol doesn't implement Hash/Eq, we compare packet contents like in original Scala code
            // which compares `_.msg.message.packet.get.content`, not the entire Protocol objects
            let request_packet_contents: HashSet<_> = requests
                .iter()
                .filter_map(|r| match &r.msg.message {
                    Some(ProtocolMessage::Packet(packet)) => Some(&packet.content),
                    _ => None,
                })
                .collect();
            let expected_packet_contents: HashSet<_> = expected_requests
                .iter()
                .filter_map(|protocol| match &protocol.message {
                    Some(ProtocolMessage::Packet(packet)) => Some(&packet.content),
                    _ => None,
                })
                .collect();
            assert_eq!(
                request_packet_contents, expected_packet_contents,
                "Request packet contents should match expected packet contents (order doesn't matter)"
            );

            fixture.transport_layer.reset();

            let last_approved_block_o = fixture.last_approved_block.lock().unwrap().clone();
            assert!(last_approved_block_o.is_some());

            {
                let engine = engine_cell.get().await;
                engine
                    .handle(
                        fixture.local.clone(),
                        CasperMessage::ApprovedBlockRequest(ApprovedBlockRequest {
                            identifier: "test".to_string(),
                            trim_state: false,
                        }),
                    )
                    .await
                    .expect("Failed to handle approved block request");
            };

            let requests_after = fixture.transport_layer.get_all_requests();
            let approved_block_bytes =
                prost::bytes::Bytes::from(approved_block.clone().to_proto().encode_to_vec());
            let found_approved_block = requests_after.iter().any(|r| match &r.msg.message {
                Some(ProtocolMessage::Packet(packet)) => packet.content == approved_block_bytes,
                _ => false,
            });
            assert!(
                found_approved_block,
                "Expected to find approved block in transport layer requests"
            );
        };

        test.await;

        Self::after_each(&fixture);
    }
}

// Creates an Initializing engine using TestFixture's shared stores and managers.
// This matches Scala's approach where InitializingSpec extends Setup and uses
// implicit vals from the Setup trait (blockStore, blockDagStorage, runtimeManager, etc.).
//
// CRITICAL: Using fixture's stores ensures genesis data exported from fixture.rspace_store
// is imported into the SAME rspace_store instance, preventing storage isolation bugs.
async fn create_initializing_engine(
    fixture: &TestFixture,
    the_init: Arc<
        dyn Fn() -> Pin<Box<dyn Future<Output = Result<(), CasperError>> + Send>> + Send + Sync,
    >,
    engine_cell: Arc<EngineCell>,
) -> Result<Arc<Initializing<TransportLayerStub>>, String> {
    // Create engine-specific channels (each Initializing instance needs its own)
    let (block_tx, block_rx) = mpsc::channel::<BlockMessage>(50);
    let (tuple_tx, tuple_rx) = mpsc::channel::<StoreItemsMessage>(50);
    let (block_processing_queue_tx, _block_processing_queue_rx) = mpsc::channel(1024);

    // Use all stores and managers from fixture (matching Scala's Setup pattern)
    Ok(Arc::new(Initializing::new(
        fixture.transport_layer.as_ref().clone(),
        fixture.rp_conf_ask.clone(),
        fixture.connections_cell.clone(),
        fixture.last_approved_block.clone(),
        fixture.block_store.clone(),
        fixture.block_dag_storage.clone(),
        fixture.deploy_storage.clone(),
        fixture.casper_buffer_storage.clone(),
        fixture.rspace_state_manager.clone(),
        block_processing_queue_tx,
        fixture.blocks_in_processing.clone(),
        fixture.casper_shard_conf.clone(),
        Some(fixture.validator_id.clone()),
        the_init,
        block_tx,
        block_rx,
        tuple_tx,
        tuple_rx,
        true,
        false,
        fixture.event_publisher.clone(),
        fixture.block_retriever.clone(),
        engine_cell.clone(),
        fixture.runtime_manager.clone(),
        fixture.estimator.clone(),
        casper::rust::heartbeat_signal::new_heartbeat_signal_ref(),
    )))
}

#[tokio::test]
async fn make_transition_to_running_once_approved_block_received() {
    InitializingSpec::make_transition_to_running_once_approved_block_received().await;
}

/// Test that verifies the fix for the race condition where a slow validator
/// misses the ApprovedBlock during genesis ceremony. The fix is to proactively
/// request the ApprovedBlock when entering Initializing state, rather than
/// waiting for it to arrive (which may never happen if it was already broadcast
/// and dropped while the node was still in GenesisValidator state).
#[tokio::test]
async fn proactively_request_approved_block_on_init() {
    use casper::rust::engine::engine::Engine;
    use models::casper::ApprovedBlockRequestProto;
    use models::routing::protocol::Message as ProtocolMessage;
    use prost::Message;

    let fixture = TestFixture::new().await;

    InitializingSpec::before_each(&fixture);

    let the_init = Arc::new(|| {
        Box::pin(async { Ok(()) }) as Pin<Box<dyn Future<Output = Result<(), CasperError>> + Send>>
    });

    let engine_cell = Arc::new(EngineCell::init());

    let initializing_engine = create_initializing_engine(&fixture, the_init, engine_cell.clone())
        .await
        .expect("Failed to create Initializing engine");

    // Clear any previous transport requests
    fixture.transport_layer.reset();
    fixture
        .transport_layer
        .set_responses(|_peer, _protocol| Ok(()));

    // Call init - this should proactively request ApprovedBlock
    initializing_engine
        .init()
        .await
        .expect("init should succeed");

    // Verify that an ApprovedBlockRequest was sent to bootstrap
    let requests = fixture.transport_layer.get_all_requests();

    // Build expected content for comparison
    let expected_proto = ApprovedBlockRequestProto {
        identifier: "".to_string(),
        trim_state: true,
    };
    let expected_content = prost::bytes::Bytes::from(expected_proto.encode_to_vec());

    assert!(
        !requests.is_empty(),
        "Initializing.init should send a request to bootstrap"
    );

    let found_approved_block_request = requests.iter().any(|req| {
        if let Some(ProtocolMessage::Packet(packet)) = &req.msg.message {
            packet.content == expected_content
        } else {
            false
        }
    });

    assert!(
        found_approved_block_request,
        "Initializing.init should send ApprovedBlockRequest. Requests sent: {:?}",
        requests.iter().map(|r| &r.msg).collect::<Vec<_>>()
    );

    InitializingSpec::after_each(&fixture);
}

#[test]
fn transition_to_initializing_invokes_init_immediately() {
    use models::casper::ApprovedBlockRequestProto;
    use models::routing::protocol::Message as ProtocolMessage;
    use prost::Message;

    std::thread::Builder::new()
        .stack_size(16 * 1024 * 1024)
        .spawn(|| {
            tokio::runtime::Runtime::new().unwrap().block_on(async {
                let fixture = TestFixture::new().await;

                InitializingSpec::before_each(&fixture);

                let init_called = Arc::new(AtomicBool::new(false));
                let init_called_ref = init_called.clone();
                let the_init = Arc::new(move || {
                    let init_called_ref = init_called_ref.clone();
                    Box::pin(async move {
                        init_called_ref.store(true, Ordering::SeqCst);
                        Ok(())
                    }) as Pin<Box<dyn Future<Output = Result<(), CasperError>> + Send>>
                });

                let engine_cell = Arc::new(EngineCell::init());
                let heartbeat_signal_ref = casper::rust::heartbeat_signal::new_heartbeat_signal_ref();

                fixture.transport_layer.reset();
                fixture
                    .transport_layer
                    .set_responses(|_peer, _protocol| Ok(()));

                transition_to_initializing(
                    &fixture.block_processing_queue_tx,
                    &fixture.blocks_in_processing,
                    &fixture.casper_shard_conf,
                    &Some(fixture.validator_id.clone()),
                    the_init,
                    true,
                    false,
                    &fixture.transport_layer,
                    &fixture.rp_conf_ask,
                    &fixture.connections_cell,
                    &fixture.last_approved_block,
                    &fixture.block_store,
                    &fixture.block_dag_storage,
                    &fixture.deploy_storage,
                    &fixture.casper_buffer_storage,
                    &fixture.rspace_state_manager,
                    fixture.event_publisher.clone(),
                    fixture.block_retriever.clone(),
                    &engine_cell,
                    &fixture.runtime_manager,
                    &fixture.estimator,
                    &heartbeat_signal_ref,
                )
                .await
                .expect("transition_to_initializing should succeed");

                assert!(
                    init_called.load(Ordering::SeqCst),
                    "transition_to_initializing should call init() immediately"
                );

                let requests = fixture.transport_layer.get_all_requests();
                let expected_proto = ApprovedBlockRequestProto {
                    identifier: "".to_string(),
                    trim_state: true,
                };
                let expected_content = prost::bytes::Bytes::from(expected_proto.encode_to_vec());

                let found_approved_block_request = requests.iter().any(|req| {
                    if let Some(ProtocolMessage::Packet(packet)) = &req.msg.message {
                        packet.content == expected_content
                    } else {
                        false
                    }
                });

                assert!(
                    found_approved_block_request,
                    "transition_to_initializing should trigger ApprovedBlockRequest via immediate init; requests: {:?}",
                    requests.iter().map(|r| &r.msg).collect::<Vec<_>>()
                );

                InitializingSpec::after_each(&fixture);
            })
        })
        .unwrap()
        .join()
        .unwrap();
}
